### PR TITLE
GPU CI

### PR DESCRIPTION
**Pull request by garrettwrong**
This PR should have the required pieces to enable GPU based Github CI using `decaf`. That should unblock some GPU development/integration opportunities.

I have been unable to solve the singles bug with cufinufft on the A100 so far. Although I am still making progress, to be pragmatic I think it would be best to get what we can working now while the bug chase it back burnered. (Unfortunately I was not able to use that time most efficiently...)

### CODE REVIEW

**Comment by garrettwrong** [comment]
Noting I should update documentation (readme) as part of this PR.

File: setup.py [hunk 1]
```diff
@@ -47,7 +47,10 @@ def read(fname):
     #   for example gpu packages which may not install for all users,
     #   or developer tools that are handy but not required for users.
     extras_require={
-        "gpu": ["pycuda", "cupy", "cufinufft==1.2"],
+        "gpu_102": ["pycuda", "cupy-cuda102", "cufinufft==1.2"],
+        "gpu_110": ["pycuda", "cupy-cuda110", "cufinufft==1.2"],
+        "gpu_111": ["pycuda", "cupy-cuda111", "cufinufft==1.2"],
+        "gpu_11x": ["pycuda", "cupy-cuda11x", "cufinufft==1.2"],
```
Comment (chris-langfield) [inline setup.py:53]: Is this wildcard syntax built into `setup.py`? 

In `installation.rst` your example has `aspire[gpu_112]`. `gpu_11x` can stand in for 112, 113, etc.?

And does that then correspond to, e.g. `cupy-cuda112`?

File: .github/workflows/workflow.yml [hunk 2]
```diff
@@ -74,3 +74,13 @@ jobs:
       run: "docs/buildsite.sh"
       shell: bash
 
+  decaf_ampere:
```
Comment (chris-langfield) [inline .github/workflows/workflow.yml:77]: curious about "ampere"

File: .github/workflows/workflow.yml [hunk 3]
```diff
@@ -74,3 +74,13 @@ jobs:
...
+    runs-on: self-hosted
+    steps:
+    - uses: actions/checkout@v2
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -e ".[dev,gpu_11x]"
```
Comment (chris-langfield) [inline .github/workflows/workflow.yml:84]: Similar to question below.. you could equally do `pip install -e ".[dev, gpu_112]"` here?

File: src/aspire/nufft/cufinufft.py [hunk 4]
```diff
@@ -95,7 +102,9 @@ def transform(self, signal):
         `(ntransforms, num_pts)`.
         """
 
-        if not (signal.dtype == self.dtype or signal.dtype == self.complex_dtype):
+        if (self._original_dtype == self.dtype) and not (
```
Comment (chris-langfield) [inline src/aspire/nufft/cufinufft.py:108]: So the purpose of this is to only raise this error if the A100 32-bit patch has *not* been activated (i.e. line 35 above) 

i.e. if we set this up as a 32 bit plan but were forced to convert to 64, and we pass in a 64 bit signal, we still want it to go ahead without complaining, even though the `_original_dtype` is 32.

**Review by chris-langfield — DISMISSED** [review]
This looks great just had a few questions

File: setup.py (hunk 1 above)
Comment (chris-langfield) [reply setup.py:53]: I see that both `cupy-cuda112` and `cupy-cuda11x` both exist as libraries. So does our `[gpu_112]` extra exist at all?

File: .github/workflows/workflow.yml (hunk 2 above)
Comment (garrettwrong) [reply .github/workflows/workflow.yml:77]: Ampere is the GPU chip architecture.

File: .github/workflows/workflow.yml (hunk 3 above)
Comment (garrettwrong) [reply .github/workflows/workflow.yml:84]: No, I don't think we should have an `112`.

File: src/aspire/nufft/cufinufft.py (hunk 4 above)
Comment (garrettwrong) [reply src/aspire/nufft/cufinufft.py:108]: Yes, that was my intent. I'll add a comment. Fair question.

File: setup.py (hunk 1 above)
Comment (garrettwrong) [reply setup.py:53]: No wildcards, just the letter `x`. You are over thinking what they did... :).

Regarding the versions I picked,  they just follow `cupy` docs, unless I made a typo 😇 .

Screen Shot 2022-11-01 at 5 47 43 PM

or https://pypi.org/project/cupy/ 

Although it is true I see a lot of packages for cupy on PyPI (yikes, way too many!)... I'm hoping supporting their main doc'd versions is fine. 

I'd prefer not to have to do this, but, it is complicated... so I'm inclined to be pragmatic.

Comment (garrettwrong) [reply setup.py:53]: Ahh I see.  The `installation.rst`  is a typo, should be 11x there. Thank you for catching.

Comment (garrettwrong) [reply setup.py:53]: I shouldn't try to write code when people in the office 😇 . Fixed.

Comment (chris-langfield) [reply setup.py:53]: I think I was overthinking and didn't want to assume it was a typo 😛

File: src/aspire/nufft/cufinufft.py (hunk 4 above)
Comment (chris-langfield) [reply src/aspire/nufft/cufinufft.py:108]: looks good

**Review by chris-langfield — APPROVED** [review]

**Comment by garrettwrong** [comment]
I think we can merge this and see how `decaf` handles the load for `develop`.  We can also link up `caf` if needed.

**Pull request merged**